File: libschema/src/bundle.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};

use super::manifest::Source;
use super::meta::Meta;
use super::metadata::Versioning;

use crate::symbol::{
    Constant, Define, EnumStruct, Enumeration, Function, MethodMap, TypeDefinition, TypeSet,
};

#[derive(Deserialize, Serialize)]
pub struct Bundle {
    /// Meta descriptor of bundle content
    pub meta: Meta,

    /// Manifest source
    pub source: Source,

    /// Strand or each individual include file
    /// With optional addon metadata for versioninig
    pub strands: HashMap<String, Strand>,

    /// Current version this bundle was last parsed from
    /// Chum bucket will continue from this commit
    pub version: Option<Versioning>,
}

#[derive(Deserialize, Serialize, Default, Clone, Debug)]
pub struct Strand {
    pub functions: HashMap<String, Function>,

    pub methodmaps: HashMap<String, MethodMap>,

    pub enumstructs: HashMap<String, EnumStruct>,

    pub constants: HashMap<String, Constant>,

    pub defines: HashMap<String, Define>,

    pub enums: HashMap<String, Enumeration>,

    pub typesets: HashMap<String, TypeSet>,

    pub typedefs: HashMap<String, TypeDefinition>,
}

File: schema/src/interfaces/bundle.ts
import { IFunction, IMethodMap, IEnumStruct, IConstant, IDefine, IEnumeration, ITypeSet, ITypeDefinition } from "./symbol";
import { Meta } from "./meta";
import { Source } from "./manifest";
import { IVersioning } from './metadata';

export interface IBundle {
    /**
     * Meta descriptor of bundle content
     */
    meta: Meta;

    /**
     * Manifest source
     */
    source: Source;

    /**
     * Strand or each individual include file
     * With optional addon metadata for versioning
     */
    strands: Record<string, IStrand>;

    /**
     * Current version this bundle was last parsed from
     * Chum bucket will continue from this commit
     */
    version: IVersioning | null;
}

export interface IStrand {
    functions: Record<string, IFunction>;

    methodmaps: Record<string, IMethodMap>;

    enumstructs: Record<string, IEnumStruct>;

    constants: Record<string, IConstant>;

    defines: Record<string, IDefine>;

    enums: Record<string, IEnumeration>;

    typesets: Record<string, ITypeSet>;

    typedefs: Record<string, ITypeDefinition>;
}
